File: 200-number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0

        rows, cols = len(grid), len(grid[0])
        visited = set()
        islands = 0
        
        def bfs(row,col):
            queue = deque([(row,col)])
            visited.add((row,col))

            while queue:
                curr_row, curr_col = queue.popleft()

                for dr, dc in [(0,1), (0,-1), (1,0), (-1,0)]:
                    nr, nc = curr_row + dr, curr_col + dc

                    if (0 <= nr < rows and 0 <= nc < cols) and grid[nr][nc] == "1" and (nr, nc) not in visited:
                        queue.append((nr, nc))
                        visited.add((nr, nc))
        
        for row in range(rows):
            for col in range(cols):
                print(row,col)
                if grid[row][col] == "1" and (row,col) not in visited:
                    bfs(row,col)
                    islands += 1

        return islands

File: 200-number-of-islands/number-of-islands.cpp
class Solution {
private:
    void dfs(vector<vector<char>>& grid, size_t row, size_t col) {
        if (grid[row][col] == '0') {
            return;
        }

        grid[row][col] = '0';
        pair<int,int> dirs[] = {make_pair(1,0), make_pair(-1,0), make_pair(0,-1), make_pair(0,1)};
        for (auto i : dirs) {
            size_t r{row+i.first}, c{col+i.second};
            if ((r >= 0 && r < grid.size()) && (c >= 0 && c < grid[0].size()) ) {
                dfs(grid,r,c);
            }
        }

    }
public:
    int numIslands(vector<vector<char>>& grid) {
        int islandCount{0};

        for (size_t i = 0; i < grid.size(); i++) {
            for (size_t j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] == '1') {
                    islandCount++;
                    dfs(grid,i,j);
                }
            }
        }

        return islandCount;
    }
};